solr/core/src/java/org/apache/solr/search/CollapsingQParserPlugin.java

```diff
@@ -524,6 +533,7 @@ public int docID() {
 
     public OrdScoreCollector(int maxDoc,
                              int segments,
+                             PrimitiveMapFactory mapFactory,
```

Comment:
I agree that we should use primitive map instead of array if we don't load enough arrays. The speed of the map put/get is sufficiently fast that I suppose it should not be visible in the overall perf of this CollapsingQParser.

But do we need to have a new factory for primitive map to switch between hash map and array? I think we could need this if we often had both cases (high load and low load of the array) AND if we had a way to switch automatically. But I suppose we don't know in advance (so we don't switch auto), and that most often the load is low.
So I suggest that we don't add the complexity of these new classes (the new util.numeric package) and we just use HPPC primitive map directly.

Reply:
That is what I thought in the first place and HPPC primitive map did gave a very good result in our performance test on large collection (number of unique values in the collapsing field is 1.2million), around 5x better in qps.

But when I write a simple single-thread benchmark test (I can add it here in this PR so you can try to tune it). 
- Number of docs are 300k
- 95% percent of queries are sparse ones so we only doing collapsing on 1% of docs.
Array approach actually do better than map (around 1.5x better). The reason for that is
- We don't now how much size we gonna need for the map, so we pay lot of cost on resizing the map.
- Map needs 2 array for storing key and value seperately, plus with above point it leads to more memory usage.

I just don't want to do a commit that can potentially slow down users. Although switching to HPPC primitive map is much easier and makes thing simple for me.
Any idea on this @dsmiley @shalinmangar @joel-bernstein

Reply:
I suggest we take inspiration from two similar places in Lucene/Solr:  `org.apache.lucene.util.DocIdSetBuilder` and `org.apache.solr.search.DocSetBuilder` which are doing essentially the same thing.  Basically, start with a memory efficient data structure, and then once we reach a threshold, switch on the fly to a larger but more efficient structure.

So start with a hash based map with a substantial initial capacity (not 4!), then upgrade to an array based map impl _instead of resizing the map_ if we exceed that capacity.  This might be done with a map impl that delegates to the real impl (hppc) with switch/resize logic.

Reply:
If some preliminary perf tests show that sometime the array version is clearly faster, then I agree we should invest in an automated switch logic. But in this case the interfaces presented in this PR do not fit this switch logic: we should be able to change the underlying implementation of the map seamlessly. This is more work.

I would like to understand better the simple benchmark test you did, because the reason for the speed/memory difference is still not obvious to me.
> 95% percent of queries are sparse ones so we only doing collapsing on 1% of docs.
Does that mean that only 1% of the 300K docs are put in the map?
So on the map implementation side we have an internal capacity of 2^12 (4K) or 2^13 (8K) for 2 arrays.
On the array implementation side we have 1 array of 300K.
For the memory the map is more efficient. For the speed, it will depend on the actual load factor and the map implementation.

> So start with a hash based map with a substantial initial capacity (not 4!), then upgrade to an array based map impl
Or maybe the reverse? This depends on the map load factor.

Reply:
Ok then I will add the benchmark class here and may be delete it in subsequence commits.